### PR TITLE
[API] Adding dicom upload

### PR DESCRIPTION
**Pull request by xlecours**
Copy of https://github.com/aces/Loris/pull/5016

This adds 
`POST /candidates/<candid>/<visit>/dicoms` to upload a dicom.
`GET /candidates/<candid>/<visit>/dicoms/<filename>/processes` to get the server processes related to that dicom upload (mri_upload)
`POST /candidates/<candid>/<visit>/dicoms/<filename>/processes` to start a mri_upload server process on that dicom.
`GET /candidates/<candid>/<visit>/dicoms/<filename>/processes/<processid>` to get the state of a process

This work is based on https://github.com/aces/Loris/pull/4244 
To see this PR changes go to https://github.com/xlecours/Loris/compare/api-module-candidates-and-subs...xlecours:adding_dicom_upload

TODO:

- [ ] Add documentation for the new endpoints.
- [ ] Make the imaging_uploader use it.
- [x] Sync FilesUploadHandler.class.inc, UploadException.class.inc with John's PR https://github.com/aces/Loris/pull/4181

### CODE REVIEW

File: src/Data/Models/MRIUploadDTO.php [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        $nullOrBool = function ($val) {
+            return is_null($val) ? null : boolval($val);
+        };
+
+        $nullOrInt = function ($val) {
+            return is_null($val) ? null : intval($val);
+        };
+
+        return new MRIUploadDTO(
+            intval($row['mri_upload_id']),
```
Comment (maltheism) [inline src/Data/Models/MRIUploadDTO.php]: @xlecours is intval needed here? The `mri_upload_id` value in the row should be an int in the database.

**Comment by xlecours** [comment]
I am waiting for https://github.com/aces/Loris/pull/6671 to get merged so I can add tests

**Comment by driusan** [comment]
This needs to go in a 0.0.4-dev, not 0.0.3

**Comment by cmadjar** [comment]
@xlecours could you rebase your branch?

**Comment by ridz1208** [comment]
@xlecours I rebased this for you here https://github.com/ridz1208/Loris/tree/adding_dicom_upload

if you want I can force push to your branch or you can simply pull it and force push it yourself.

**Comment by ridz1208** [comment]
@cmadjar @driusan this one is now rebased and seems to be passing tests

**Comment by xlecours** [comment]
This is missing integration tests. @xlecours

**Comment by driusan** [comment]
@cmadjar can you review?

**Comment by xlecours** [comment]
@xlecours If you do the rebase, @cmadjar agreed to review it again. Also, now that we have a testsuite for the api module, you should add some tests.

**Comment by driusan** [comment]
@xlecours are you referring to yourself in the third person? Who's supposed to be writing tests?

**Comment by xlecours** [comment]
> @xlecours are you referring to yourself in the third person? Who's supposed to be writing tests?

Yes, I am.

**Comment by driusan** [comment]
Replaced by #9154

**Pull request closed without merging**